### PR TITLE
New maps functionality

### PR DESCRIPTION
**Pull request by vlach1989**
- new maps actions/selectors/reducers
- tests

### CODE REVIEW

File: src/state/Maps/actions.js [hunk 1]
```diff
@@ -370,6 +473,74 @@ function setMapSetBackgroundLayer(setKey, backgroundLayer) {
...
+				dispatch(layerUse(backgroundLayerState, spatialFilter));
+			} else {
+				return;
+			}
+		}
+	};
+}
+
+/**
+ * TODO @vlach1989 test
```
Comment (vlach1989) [inline src/state/Maps/actions.js:513]:
```suggestion
```

File: src/state/Maps/selectors.js [hunk 2]
```diff
@@ -36,6 +37,17 @@ const getActiveMapKey = state => state.maps.activeMapKey;
 const getMapsAsObject = state => state.maps.maps;
 const getMapSetsAsObject = state => state.maps.sets;
 
+const getActiveMap = createSelector(
+	[getMapsAsObject, getActiveMapKey],
+	(maps, activeMapKey) => {
+		if (maps && activeMapKey) {
+			return maps[activeMapKey];
```
Comment (vlach1989) [inline src/state/Maps/selectors.js:44]:
```suggestion
			return maps[activeMapKey] || null;
```

File: src/state/Maps/reducers.js [hunk 3]
```diff
@@ -155,6 +335,56 @@ const setSetBackgroundLayer = (state, setKey, backgroundLayer) => {
...
+ * @return {Object} state
+ */
+const setSetLayers = (state, setKey, layers) => {
+	if (setKey && layers && state.sets[setKey]) {
+		return {
+			...state,
+			sets: {
+				...state.sets,
+				[setKey]: {
+					...state.sets[setKey],
```
Comment (vdubr) [inline src/state/Maps/reducers.js:375]: Could be undefined.
But, I dont know if is right approach to create new one.

File: src/state/Maps/selectors.js [hunk 4]
```diff
@@ -654,18 +689,26 @@ const getFinalLayerByDataSourceAndLayerState = createRecomputeSelector(
...
 					? configuration.singleTile
 					: false;
 
 			const styles = rest.styles;
 
+			// if path is relative, add protocol & host from config
+			// TODO special config constants for mapserver?
+			if (url && url.charAt(0) !== 'h') {
+				const localConfig = AppSelectors.getCompleteLocalConfigurationObserver();
+				url = `${localConfig.apiBackendProtocol}://${localConfig.apiBackendHost}/${url}`;
```
Comment (vdubr) [inline src/state/Maps/selectors.js:704]: We should move it to the helper which identify relative URL and fix it to the absolute

File: src/state/Maps/reducers.js [hunk 5]
```diff
@@ -40,6 +65,116 @@ const removeMapFromSet = (state, setKey, mapKey) => {
...
+const addMapLayers = (state, mapKey, layerStates) => {
+	if (mapKey && layerStates && state.maps?.[mapKey]) {
+		return {
+			...state,
+			maps: {
+				...state.maps,
+				[mapKey]: {
+					...state.maps[mapKey],
+					data: {
+						...state.maps[mapKey].data,
```
[the remark this answers is not in the document]
Comment (vlach1989) [reply src/state/Maps/reducers.js:121]: It couldn't be undefined because of line 113

File: src/state/Maps/selectors.js (hunk 4 above)
Comment (vdubr) [reply src/state/Maps/selectors.js:704]: in future

**Review by vdubr — APPROVED** [review]

File: src/state/Maps/reducers.js [hunk 6]
```diff
@@ -81,6 +223,51 @@ const setMapLayerStyleKey = (state, mapKey, layerKey, styleKey) => {
...
+ * Set map layer option
+ * @param state {Object}
+ * @param mapKey {string}
+ * @param layerKey {string}
+ * @param optionKey {string}
+ * @param optionValue {*}
+ * @return {Object} state
+ */
+const setMapLayerOption = (state, mapKey, layerKey, optionKey, optionValue) => {
+	const layers = state.maps[mapKey]?.data?.layers;
```
Comment (tmdmc) [inline src/state/Maps/reducers.js:236]: Should we check for mapKey, too?

File: src/state/Maps/selectors.js [hunk 7]
```diff
@@ -654,18 +689,26 @@ const getFinalLayerByDataSourceAndLayerState = createRecomputeSelector(
...
 			const singleTile =
 				configuration && configuration.hasOwnProperty('singleTile')
 					? configuration.singleTile
 					: false;
 
 			const styles = rest.styles;
 
+			// if path is relative, add protocol & host from config
+			// TODO special config constants for mapserver?
+			if (url && url.charAt(0) !== 'h') {
```
Comment (tmdmc) [inline src/state/Maps/selectors.js:702]: Ah, the old Windows 9 killer. Could be more robust, or at least explained in comment.

**Review by tmdmc — CHANGES_REQUESTED** [review]

File: src/state/Maps/reducers.js (hunk 6 above)
Comment (vlach1989) [reply src/state/Maps/reducers.js:236]: IMO we don't need it here. You are right that in some reducers is the check redundant

File: src/state/Maps/selectors.js (hunk 7 above)
Comment (vlach1989) [reply src/state/Maps/selectors.js:703]: https://github.com/gisat-panther/ptr-state/pull/77/commits/46d967317f429895c2fb50b76f565e405fcfd2a6
Comment added. Check of valid URL is not trivial. Moreover, we need to discuss all the possibilities that are relevant to us as a url parameter

**Pull request merged**